drivers/sensor/bma4xx/bma4xx_emul.c

```diff
@@ -0,0 +1,428 @@
+	int acceleration[3];
+	/** Axis offset register values, XYZ. */
+	uint8_t offset[3];
+	/** NV_CONF register value. */
+	uint8_t nv_config;
+	/**
+	 * True if the sensor FIFO is currently enabled.
+	 *
+	 * Only headerless mode for the accelerometer alone is supported.
+	 */
+	bool fifo_enabled;
+	/** Pointer to data that will be read from the FIFO. */
+	const uint8_t *fifo_data;
+	/** Number of bytes remaining in fifo_data. */
+	uint16_t fifo_available;
+
+	/** True if in latched interrupt mode, otherwise non-latched. */
+	bool interrupt_mode_latched;
+	/** Raw value of INT1_IO_CTRL register. */
+	uint8_t int1_io_ctrl;
+	/** Raw value of INT_MAP_DATA register. */
+	uint8_t int_map_data;
+};
+
+struct bma4xx_emul_cfg {
+};
+
+void bma4xx_emul_reset(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	data->regs[BMA4XX_REG_CHIP_ID] = BMA4XX_CHIP_ID;
+	data->accel_enabled = false;
+	data->accel_range = 2;
+	data->odr_raw = 8;
+	memset(data->acceleration, 0, sizeof(data->acceleration));
+	memset(data->offset, 0, sizeof(data->offset));
+	data->nv_config = 0;
+	data->fifo_enabled = false;
+	data->fifo_data = NULL;
+	data->fifo_available = 0;
+	data->interrupt_mode_latched = false;
+}
+
+void bma4xx_emul_set_reg(const struct emul *target, uint8_t reg_addr, const uint8_t *val,
+			 size_t count)
+{
+	struct bma4xx_emul_data *data = target->data;
+
+	__ASSERT_NO_MSG(reg_addr + count <= NUM_REGS);
+	memcpy(data->regs + reg_addr, val, count);
+}
+
+void bma4xx_emul_get_reg(const struct emul *target, uint8_t reg_addr, uint8_t *val, size_t count)
+{
+	struct bma4xx_emul_data *data = target->data;
+
+	__ASSERT_NO_MSG(reg_addr + count <= NUM_REGS);
+	memcpy(val, data->regs + reg_addr, count);
+}
+
+bool bma4xx_emul_is_accel_enabled(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	return data->accel_enabled;
+}
+
+uint8_t bma4xx_emul_get_accel_range(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	return data->accel_range;
+}
+
+uint32_t bma4xx_emul_get_odr(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	/*
+	 * This function deliberately differs from bma4xx_reg_to_odr() to provide
+	 * an obviously-correct reference for tests.
+	 */
+	switch (data->odr_raw) {
+	case 1:
+		return 781; /* 25/32 Hz */
+	case 2:
+		return 1562; /* 25/16 Hz */
+	case 3:
+		return 3125;
+	case 4:
+		return 6250;
+	case 5:
+		return 12500;
+	case 6:
+		return 25000;
+	case 7:
+		return 50000;
+	case 8:
+		return 100000;
+	case 9:
+		return 200000;
+	case 10:
+		return 400000;
+	case 11:
+		return 800000;
+	case 12:
+		return 1600000;
+	default:
+		LOG_ERR("ODR register value %#x is reserved", data->odr_raw);
+		return 0;
+	}
+}
+
+void bma4xx_emul_set_accel_data(const struct emul *emul, int x, int y, int z)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	data->acceleration[0] = x;
+	data->acceleration[1] = y;
+	data->acceleration[2] = z;
+}
```

Comment:
I think @tristan-google commented on this (lets sync later today) this should be done via the backend channel set.